sample/samplers.go

```diff
@@ -2,43 +2,88 @@ package sample
 
 import (
 	"errors"
+	"math"
 	"math/rand/v2"
 	"slices"
-)
+	"sync"
 
-// Sampler is not thread-safe. Each goroutine should have its own instance
-type Sampler interface {
-	Sample([]float32) (int32, error)
-}
+	"github.com/ollama/ollama/llama"
+)
 
-// logit represents information about a single token during sampling
-type logit struct {
+// token represents information about a single token during sampling
+type token struct {
 	id    int32   // The token's unique identifier
 	value float32 // The raw logit or probability from the model
 }
 
-type weighted struct {
+type Sampler struct {
 	rng         *rand.Rand
-	tokens      []logit
 	topK        int
 	topP        float32
 	minP        float32
 	temperature float32
+	grammar     *Grammar
 }
 
-func (s *weighted) Sample(logits []float32) (int32, error) {
-	if len(s.tokens) < len(logits) {
-		s.tokens = make([]logit, len(logits))
+func (s *Sampler) Sample(logits []float32) (int32, error) {
+	tokens := make([]token, len(logits))
+	for i := range logits {
+		tokens[i].id = int32(i)
+		tokens[i].value = logits[i]
 	}
 
-	tokens := s.tokens[:len(logits)]
+	t, err := s.sample(tokens)
+	if err != nil {
+		return -1, err
+	}
```

Comment:
We should be doing the grammar check earlier so that we can greedy sample directly instead of running through the entire sampler chain. Maybe something like this:
```suggestion
		var t token
	var err error
	if s.grammar != nil {
		t = s.greedy(tokens)
	} else {
		t, err = s.sample(tokens)
		if err != nil {
			return -1, err
		}
	}
```

Reply:
Wouldn't we lose the combo of having weighted sampling + grammar constraining in that case? LMK if that's the behavior we want and I can update

Reply:
I don't think so as weighted sampling should only get called in the next `if` block if the greedy token was invalid in the grammar